cli/cmd/config.go

```diff
@@ -70,138 +200,8 @@ func setupCommand(options ...setupOption) func(cmd *cobra.Command, args []string
-						{Header: "ID", Path: "spec.id"},
-						{Header: "NAME", Path: "spec.name"},
-						{Header: "STRATEGY", Path: "spec.strategy"},
-					},
-				},
-			),
-		)
-
-		resources.Register(
-			resourcemanager.NewClient(
-				httpClient,
-				"demo", "demos",
-				resourcemanager.TableConfig{
-					Cells: []resourcemanager.TableCellConfig{
-						{Header: "ID", Path: "spec.id"},
-						{Header: "NAME", Path: "spec.name"},
-						{Header: "TYPE", Path: "spec.type"},
-						{Header: "ENABLED", Path: "spec.enabled"},
-					},
-				},
-			),
-		)
-
-		resources.Register(
-			resourcemanager.NewClient(
-				httpClient,
-				"datastore", "datastores",
-				resourcemanager.TableConfig{
-					Cells: []resourcemanager.TableCellConfig{
-						{Header: "ID", Path: "spec.id"},
-						{Header: "NAME", Path: "spec.name"},
-						{Header: "DEFAULT", Path: "spec.default"},
-					},
-					ItemModifier: func(item *gabs.Container) {
-						isDefault := item.Path("spec.default").Data().(bool)
-						if !isDefault {
-							item.SetP("", "spec.default")
-						} else {
-							item.SetP("*", "spec.default")
-						}
-					},
-				},
-			),
-		)
-
-		resources.Register(
-			resourcemanager.NewClient(
-				httpClient,
-				"environment", "environments",
-				resourcemanager.TableConfig{
-					Cells: []resourcemanager.TableCellConfig{
-						{Header: "ID", Path: "spec.id"},
-						{Header: "NAME", Path: "spec.name"},
-						{Header: "DESCRIPTION", Path: "spec.description"},
-					},
-				},
-			),
-		)
-
-		resources.Register(
-			resourcemanager.NewClient(
-				httpClient,
-				"transaction", "transactions",
-				resourcemanager.TableConfig{
-					Cells: []resourcemanager.TableCellConfig{
-						{Header: "ID", Path: "spec.id"},
-						{Header: "NAME", Path: "spec.name"},
-						{Header: "VERSION", Path: "spec.version"},
-						{Header: "STEPS", Path: "spec.summary.steps"},
-						{Header: "RUNS", Path: "spec.summary.runs"},
-						{Header: "LAST RUN TIME", Path: "spec.summary.lastRun.time"},
-						{Header: "LAST RUN SUCCESSES", Path: "spec.summary.lastRun.passes"},
-						{Header: "LAST RUN FAILURES", Path: "spec.summary.lastRun.fails"},
-					},
-					ItemModifier: func(item *gabs.Container) {
-						// set spec.summary.steps to the number of steps in the transaction
-						item.SetP(len(item.Path("spec.steps").Children()), "spec.summary.steps")
-
-						// if lastRun.time is not empty, show it in a nicer format
-						lastRunTime := item.Path("spec.summary.lastRun.time").Data().(string)
-						if lastRunTime != "" {
-							date, err := time.Parse(time.RFC3339, lastRunTime)
-							if err != nil {
-								panic(err)
-							}
-							if date.IsZero() {
-								item.SetP("", "spec.summary.lastRun.time")
-							} else {
-								item.SetP(date.Format(time.DateTime), "spec.summary.lastRun.time")
-							}
-						}
-					},
-				},
-			),
-		)
+		hc := resourcemanager.NewHTTPClient(cliConfig.URL(), extraHeaders)
+		*httpClient = *hc
```

Comment:
Question: is there a reason to do this pointer trick instead of doing something like this:
```suggestion
		httpClient = &resourcemanager.NewHTTPClient(cliConfig.URL(), extraHeaders)
```

Reply:
Yes. In order to have this dynamic list available, we need to register all resources before the commands are registered. To do this, the resources are registered as a package variable [here](https://github.com/kubeshop/tracetest/pull/2832/files#diff-7e99687b5d846708c7b5a928628f9646f5c6f4bd52c231a75a46b78b09ca57adR53), and the commands are registered within `init` functions. This guarantees that the registry will be ready before the commands.

The problem with this is that the resources depened on the `resourcemanager.HTTPClient`, which in turns depends on the config being ready. The config gets ready in the `setupCommand` function, which is called as a `preRun` of the commands. So in order to configure a command, you need to have already configured the command. This is a ciruclar reference initialization, and that's not allowed.

The solution I found was to setup the registry with a pointer to an empty `HTTPClient`, and when the `setupCommand` runs on the `preRun` stage of the command, it replace the pointer with the correctly configured client.

Does this make sense?

Reply:
Yes, thanks!

Reply:
Just one detail: can we add a small comment on the code explaining that? Maybe just a link to you explanation on the PR

Reply:
Good idea. Added!